lib/RateLimiterSQLite.js

```diff
@@ -0,0 +1,214 @@
+      this.client.run(createTableSQL, (err) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve();
+        }
+      });
+    });
+  }
+
+  _getCreateTableSQL() {
+    return `CREATE TABLE IF NOT EXISTS ${this.tableName} (
+      key TEXT PRIMARY KEY,
+      points INTEGER NOT NULL DEFAULT 0,
+      expire INTEGER
+    )`;
+  }
+
+  _clearExpiredHourAgo() {
+    if (this._clearExpiredTimeoutId) {
+      clearTimeout(this._clearExpiredTimeoutId);
+    }
+    this._clearExpiredTimeoutId = setTimeout(() => {
+      this.clearExpired(Date.now() - 3600000) // Never rejected
+        .then(() => {
+          this._clearExpiredHourAgo();
+        });
+    }, 300000);
+    this._clearExpiredTimeoutId.unref();
+  }
+
+  clearExpired(nowMs) {
+    return new Promise((resolve) => {
+      this.client.run(
+        `DELETE FROM ${this.tableName} WHERE expire < ?`,
+        [nowMs],
+        () => resolve()
+      );
+    });
+  }
+
+  _getRateLimiterRes(rlKey, changedPoints, result) {
+    const res = new RateLimiterRes();
+    res.isFirstInDuration = changedPoints === result.points;
+    res.consumedPoints = res.isFirstInDuration ? changedPoints : result.points;
+    res.remainingPoints = Math.max(this.points - res.consumedPoints, 0);
+    res.msBeforeNext = result.expire
+      ? Math.max(result.expire - Date.now(), 0)
+      : -1;
+
+    return res;
+  }
+
+  _upsert(rlKey, points, msDuration, forceExpire = false) {
+    if (!this.tableCreated) {
+      return Promise.reject(Error("Table is not created yet"));
+    }
+
+    return new Promise((resolve, reject) => {
+      const dateNow = Date.now();
+      const newExpire = msDuration > 0 ? dateNow + msDuration : null;
+
+      this.client.serialize(() => {
+        if (forceExpire) {
+          this.client.run(
+            `INSERT OR REPLACE INTO ${this.tableName} (key, points, expire) VALUES (?, ?, ?)`,
+            [rlKey, points, newExpire],
+            (err) => {
+              if (err) {
+                reject(err);
+              } else {
+                resolve({ points, expire: newExpire });
+              }
+            }
+          );
+        } else {
+          // Increment points if exists and not expired, otherwise insert
+          this.client.run(
+            `
+              INSERT INTO ${this.tableName} (key, points, expire)
+              VALUES (?, ?, ?)
+              ON CONFLICT(key) DO UPDATE SET
+                points =  CASE
+                  WHEN expire is NULL OR expire > ? THEN points + excluded.points
+                  ELSE excluded.points
+                END,
+                expire = CASE
+                  WHEN expire is NULL OR expire > ? THEN expire
+                  ELSE excluded.expire
+                END
+
+            `,
+            [rlKey, points, newExpire, dateNow, dateNow],
+            (err) => {
+              if (err) {
+                reject(err);
+              }
+
+              this.client.get(
```

Comment:
@muco-rolle This query should be run in transaction with the upsert query. The `_upsert` function should start a transaction, make two above queries and commit transaction. If there is an error on any step it should rollback transaction.
One good example is MySQL limiter https://github.com/animir/node-rate-limiter-flexible/blob/0a2a27df6be4027c7350d88d1f19dfde287d08eb/lib/RateLimiterMySQL.js